src/vmm/src/device_manager/mmio.rs

```diff
@@ -74,43 +76,65 @@ pub struct MMIODeviceInfo {
     pub irqs: Vec<u32>,
 }
 
+impl Aml for MMIODeviceInfo {
+    fn append_aml_bytes(&self, v: &mut Vec<u8>) {
+        let dev_id = self.irqs[0] - arch::IRQ_BASE;
+        debug!(
+            "acpi: Building AML for VirtIO device _SB_.V{:03}. memory range: {:#010x}:{} irq: {}",
+            dev_id, self.addr, self.len, self.irqs[0]
+        );
+        aml::Device::new(
+            format!("V{:03}", dev_id).as_str().into(),
+            vec![
+                &aml::Name::new("_HID".into(), &"LNRO0005"),
+                &aml::Name::new("_UID".into(), &dev_id),
+                &aml::Name::new("_CCA".into(), &aml::ONE),
+                &aml::Name::new(
+                    "_CRS".into(),
+                    &aml::ResourceTemplate::new(vec![
+                        &aml::Memory32Fixed::new(
+                            true,
+                            self.addr.try_into().unwrap(),
+                            self.len.try_into().unwrap(),
+                        ),
+                        &aml::Interrupt::new(true, true, false, false, self.irqs[0]),
+                    ]),
+                ),
+            ],
+        )
+        .append_aml_bytes(v);
+    }
+}
+
 /// Manages the complexities of registering a MMIO device.
 #[derive(Debug)]
 pub struct MMIODeviceManager {
     pub(crate) bus: crate::devices::Bus,
-    pub(crate) irq_allocator: IdAllocator,
-    pub(crate) address_allocator: AddressAllocator,
+    pub(crate) resource_allocator: Rc<ResourceAllocator>,
     pub(crate) id_to_dev_info: HashMap<(DeviceType, String), MMIODeviceInfo>,
+    pub(crate) dsdt_data: Vec<u8>,
```

Comment:
That's how I had it initially. The problem with this, is that iterating over the bus afterwards won't give me the devices in the order they were added in the bus. That is problem, because the root file system disk needs to appear first in the AML, so that the guest sees it as `/dev/vda`.

Reply:
Ah, I see, that's unfortunate. I guess we could use an `indexmap` instead of a `HashMap`, but that'd mean another dependency, so probably not worth it. Could we leave a comment on this field explaining what you explained in this comment? :o